### PR TITLE
Track Firefox downloads from /thanks in GA4

### PR DESCRIPTION
**Pull request by stephaniehobson**
## One-line summary

Triggers `product_download` from the /thanks page to record the download in GA4.

## Significant changes and points to review

- convert thanks.js and thanks-direct.js to ES6
- add listener to the direct download link that displays as a fallback
- trigger product_download event when automatic download triggers

## Issue / Bugzilla link

#13238 

## Testing

[Integration test pass](https://github.com/mozilla/bedrock/actions/runs/5615441318)

I avoided downloading a million files by changing my browser settings so it asks where to save the file before starting the download, which then also gives the option to cancel the download before it begins.

This branch is up on demo4 for testing with https://tagassistant.google.com/

https://www-demo4.allizom.org/en-US/firefox/new/ (and click download)
https://www-demo4.allizom.org/en-US/firefox/download/thanks/?s=direct

### CODE REVIEW

**Comment by stephaniehobson** [comment]
Needs support for firefox-stub and firefox-msi download URLs.

File: media/js/firefox/new/common/thanks-direct.es6.js [hunk 1]
```diff
@@ -4,17 +4,17 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-(function () {
-    'use strict';
+import TrackProductDownload from '../../../base/datalayer-productdownload.es6';
 
-    var timeout;
-    var requestComplete = false;
+(function () {
```
Comment (alexgibson) [inline media/js/firefox/new/common/thanks-direct.es6.js:9]: NIt: no need to wrap this in an IIFE, webpack will do it for you

File: media/js/firefox/new/common/thanks-direct.es6.js [hunk 2]
```diff
@@ -31,10 +31,25 @@
...
+                    directDownloadLink.addEventListener(
+                        'click',
+                        function (event) {
+                            try {
+                                TrackProductDownload.handleLink(event);
+                            } catch (error) {
+                                return;
+                            }
+                        }
+                    );
```
Comment (alexgibson) [inline media/js/firefox/new/common/thanks-direct.es6.js:43]: I think you can simplify this a bit by using an arrow function

(also make sure to add the `useCapture` param at the end for legacy browser support?):

```suggestion
                    directDownloadLink.addEventListener('click', (event) => {
                        try {
                            TrackProductDownload.handleLink(event);
                        } catch (error) {
                            return;
                        }
                    }, false);
```

File: media/js/firefox/new/common/thanks-init.es6.js [hunk 3]
```diff
@@ -4,11 +4,11 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/.
  */
 
-(function () {
-    'use strict';
+import TrackProductDownload from '../../../base/datalayer-productdownload.es6';
 
-    var directDownloadLink = document.getElementById('direct-download-link');
-    var downloadURL;
+(function () {
```
Comment (alexgibson) [inline media/js/firefox/new/common/thanks-init.es6.js:9]: Nit: No IIFE needed

File: media/js/firefox/new/common/thanks-init.es6.js [hunk 4]
```diff
@@ -25,10 +25,22 @@
             // Make sure the 'Try downloading again' link is well formatted! (issue 9615)
             if (directDownloadLink && directDownloadLink.href) {
                 directDownloadLink.href = downloadURL;
+                directDownloadLink.addEventListener('click', function (event) {
```
Comment (alexgibson) [inline media/js/firefox/new/common/thanks-init.es6.js:28]: Same as above, simplify to arrow function and add `false` for `useCapture`

**Review by alexgibson — CHANGES_REQUESTED** [review]
Code changes look good, just a few minor nits / suggestions

I do have a larger question though: would it make more sense to add this as a global script that can track downloads on button click, as opposed to when loading the /thanks/ page? I know Adria has talked about wanting to move in this direction for how we measure things.

Right now, this won't capture Firefox downloads for pre-release channels, ESR downloads links (which can now appear on many pages), or from other pages like installer-help where we do offer direct download links.

If there's a chance that this might end up as a global script as you make more progress, might it be worth just making it global now and triggering on button click?

**Comment by alexgibson** [comment]
> I do have a larger question though: would it make more sense to add this as a global script that can track downloads on button click, as opposed to when loading the /thanks/ page? I know Adria has talked about wanting to move in this direction for how we measure things.

> Right now, this won't capture Firefox downloads for pre-release channels, ESR downloads links (which can now appear on many pages), or from other pages like installer-help where we do offer direct download links.

> If there's a chance that this might end up as a global script as you make more progress, might it be worth just making it global now and triggering on button click?

I've given this some further thought, and realized you've probably done this so we can still track downloads that link directly to the /thanks/ page (e.g. from AMO?) With that in mind, I think what you have here makes sense for now.

I do wonder if the script needs to be globally available in the base bundle though, which is what you seem to be moving toward in the adjust PR?

**Comment by stephaniehobson** [comment]
> I do wonder if the script needs to be globally available in the base bundle though, which is what you seem to be moving toward in the adjust PR?

Yup! I'm just rolling it out in bits to make the reviews easier (though, now I see this makes reviewing the overall approach harder :/ )

> Noticed a possible bug: when I download on windows, release_channel has the value of stub, but when on mac, the value is release?

This should be fixed in the adjust PR.

**Review by alexgibson — APPROVED** [review]
Code changes look good to me r+

As you're now transpiling this code and it's a pretty critical page, I wonder if it's worth pushing this to demo and at least testing that the download still triggers like expected on some older browsers. Thoughts?

Alternatively, you could add your tracking code to the `data` bundle globally now (instead of the next PR that will presumably revert these changes?).

**Comment by stephaniehobson** [comment]
> Alternatively, you could add your tracking code to the `data` bundle globally now (instead of the next PR that will presumably revert these changes?).

Hm, I didn't consider that I might be able to use TrackProductDownload without importing it 🤔

**Comment by stephaniehobson** [comment]
Going to combine with #13416.

**Pull request closed without merging**